compiler-core/src/javascript/typescript.rs

```diff
@@ -0,0 +1,530 @@
+            Statement::TypeAlias { .. } => vec![],
+
+            // TODO: Do we need to do anything here?
+            Statement::ExternalType { .. } => vec![],
+
+            Statement::Import { .. } => vec![],
+
+            Statement::CustomType {
+                public,
+                constructors,
+                opaque,
+                name,
+                ..
+            } if *public && !*opaque => self.custom_type_definition(name, constructors),
+            Statement::CustomType { .. } => vec![],
+
+            Statement::ModuleConstant {
+                public,
+                name,
+                value,
+                ..
+            } if *public => vec![self.module_constant(name, value)],
+            Statement::ModuleConstant { .. } => vec![],
+
+            Statement::Fn {
+                arguments,
+                name,
+                public,
+                return_type,
+                ..
+            } if *public => vec![self.module_function(name, arguments, return_type)],
+            Statement::Fn { .. } => vec![],
+
+            Statement::ExternalFn {
+                public,
+                name,
+                arguments,
+                return_type,
+                ..
+            } if *public => vec![self.external_function(name, arguments, return_type)],
+            Statement::ExternalFn { .. } => vec![],
+        }
+    }
+
+    fn type_alias(&self, alias: String, type_: &Type) -> Output<'a> {
+        Ok(docvec![
+            "export type ",
+            Document::String(ts_safe_type_name(alias)),
+            " = ",
+            self.type_printer.print(type_),
+            ";"
+        ])
+    }
+
+    /// Converts a Gleam custom type definition into the Typescript equivalent.
+    /// In Gleam, all custom types have one to many concrete constructors. This
+    /// function first converts the constructors into Typescript then finally
+    /// emits a union type to represent the Typescript type itself. Because in
+    /// Gleam constructors can have the same name as the custom type, here we
+    /// append a "$" symbol to the emited Typescript type to prevent those
+    /// naming clases.
+    ///
+    fn custom_type_definition(
+        &mut self,
+        name: &str,
+        constructors: &'a [TypedRecordConstructor],
+    ) -> Vec<Output<'a>> {
+        let mut definitions: Vec<Output<'_>> = constructors
+            .iter()
+            .map(|constructor| Ok(self.record_definition(constructor)))
+            .collect();
+
+        definitions.push(Ok(docvec![
+            Document::String("export type ".to_string()),
+            Document::String(format!("{}$", name)),
+            Document::String(" = ".to_string()),
+            concat(Itertools::intersperse(
+                constructors.iter().map(|x| x.name.to_doc()),
+                break_("| ", " | "),
+            )),
+            Document::String(";".to_string()),
```

Comment:
```suggestion
            "export type ",
            name,
            "$ = ",
            concat(Itertools::intersperse(
                constructors.iter().map(|x| x.name.to_doc()),
                break_("| ", " | "),
            )),
            ";",
```

Reply:
I was getting some weird errors from Rust about not using a variable where it was defined without this (I think it shows up in the second video), but I'll try again without these!